src/java/com/puppetlabs/puppetserver/MetricsPuppetProfiler.java

```diff
@@ -54,29 +62,59 @@ public Set<String> getAllMetricIds() {
     }
 
     public Map<String, Timer> getFunctionTimers() {
-        return getTimers(FUNCTION_PATTERN);
+        return this.function_timers;
     }
 
     public Map<String, Timer> getResourceTimers() {
-        return getTimers(RESOURCE_PATTERN);
+        return this.resource_timers;
     }
 
     public Map<String, Timer> getCatalogTimers() {
-        return getTimers(CATALOG_PATTERN);
+        return this.catalog_timers;
     }
 
     public Map<String, Timer> getInliningTimers() {
-        return getTimers(INLINING_PATTERN);
+        return this.inlining_timers;
     }
 
     public Map<String, Timer> getPuppetDBTimers() {
-        return getTimers(PUPPETDB_PATTERN);
+        return this.puppetdb_timers;
     }
 
     @Override
     public void shutdown() {
     }
 
+    private List<String> sliceOfArrayToList(String[] idSegments, int lengthOfID) {
+        // Callers expect a mutable List returned, but Arrays.asList() returns a
+	// fix length array, which is why we have to create a List and then add to it.
+        List<String> idList = new ArrayList<String>();
+        idList.addAll(Arrays.asList(Arrays.copyOf(idSegments, lengthOfID)));
+
+	return idList;
+    }
+
+    private void updateMetricsTrackers(String[] metricId, Map<String, Timer> metricsByID) {
+        if ("functions".equals(metricId[0])) {
+            Timer metric = metricsByID.get(getMetricName(sliceOfArrayToList(metricId, 2)));
+            this.function_timers.put(metricId[1], metric);
+        } else if ("compiler".equals(metricId[0])) {
+            if ("evaluate_resource".equals(metricId[1])) {
+                Timer metric = metricsByID.get(getMetricName(sliceOfArrayToList(metricId, 3)));
+                this.resource_timers.put(metricId[2], metric);
+            } else if ("static_compile_inlining".equals(metricId[1])) {
+                Timer metric = metricsByID.get(getMetricName(sliceOfArrayToList(metricId, 3)));
+                this.inlining_timers.put(metricId[2], metric);
+            } else {
+                Timer metric = metricsByID.get(getMetricName(sliceOfArrayToList(metricId, 2)));
+                this.catalog_timers.put(metricId[1], metric);
+            }
+        } else if ("puppetdb".equals(metricId[0])) {
```

Comment:
This puppetdb clause does not contain all the necessary complexity of the former PUPPETDB_PATTERN

Reply:
These are the root metrics we report on based the PUPPETDB_PATTERN:

```
puppetdb.catalog.munge
puppetdb.command.submit.replace facts
puppetdb.command.submit.replace catalog
puppetdb.command.submit.store report
puppetdb.facts.encode
puppetdb.query
puppetdb.resource.search
puppetdb.report.convert_to_wire_format_hash
```

And, going through the PuppetDB Ruby code (at this commit https://github.com/puppetlabs/puppetdb/commit/8a1e2efe2c608c11453fae689a45150c0154aea1) I see these metrics created:
```
[:puppetdb, :aliases, :map_to_title]
[:puppetdb, :catalog, :munge]
[:puppetdb, :catalog, :save, request.key]
[:puppetdb, :command, :submit]
[:puppetdb, :command, :submit, catalog_command_name, catalog_version]
[:puppetdb, :command, :submit, command_name, version]
[:puppetdb, :command, :submit, inputs_command_name, inputs_version]
[:puppetdb, :edges, :munge]
[:puppetdb, :edges, :synthesize]
[:puppetdb, :edges, :synthesize, :make_unique]
[:puppetdb, :edges, :synthesize, :primary_synthesis]
[:puppetdb, :edges, :synthesize, :resource_table, :build]
[:puppetdb, :events_list, :build]
[:puppetdb, :facts, :encode]
[:puppetdb, :facts, :find, :parse_response, request.key]
[:puppetdb, :facts, :find, :query_nodes, request.key]
[:puppetdb, :facts, :find, request.key]
[:puppetdb, :facts, :save, request.key]
[:puppetdb, :facts, :search, :parse_query_response, request.key,]
[:puppetdb, :facts, :search, :query_request, request.key]
[:puppetdb, :facts, :search, request.key]
[:puppetdb, :keys, :filter_extraneous]
[:puppetdb, :logs_list, :build]
[:puppetdb, :metaparams, :sort]
[:puppetdb, :metrics_list, :build]
[:puppetdb, :namevar_aliases, :add]
[:puppetdb, :parameters, :add_missing]
[:puppetdb, :payload, :format]
[:puppetdb, :query, query]
[:puppetdb, :report, :convert_to_wire_format_hash]
[:puppetdb, :report, :process]
[:puppetdb, :resource, :search, :build_up_collected_objects, request.key]
[:puppetdb, :resource, :search, :parse_query_response, request.key]
[:puppetdb, :resource, :search, :query, request.key]
[:puppetdb, :resource, :search, request.key]
[:puppetdb, :resources_list, :build]
[:puppetdb, :titles, :stringify]
```

Are the current metrics correct? We could add a few while we're here. Some easier ones to add would be:
```
puppetdb.catalog.save
puppetdb.edges.munge
puppetdb.edges.synthesize
puppetdb.facts.find
puppetdb.facts.save
puppetdb.facts.search
puppetdb.payload.format
puppetdb.report.process
```
But I don't know if they duplicate some existing reported metrics, or aren't very valuable.

Thoughts @austb, @rbrw , @Sharpie ?

Reply:
@Sharpie , how does adding those new puppetdb metrics sound?

Reply:
Here's the updated metrics:
```
      "puppetdb-metrics": [
        {                                         
          "metric": "catalog_save",
          "count": 2,
          "mean": 243,   
          "aggregate": 486
        },
        {                            
          "metric": "resource_search",
          "count": 8,
          "mean": 46,    
          "aggregate": 368
        },
        {                                                                                                
          "metric": "query",
          "count": 1,
          "mean": 355,  
          "aggregate": 355
        },
        {                          
          "metric": "facts_find",
          "count": 2,
          "mean": 117,  
          "aggregate": 234
        },
        {                 
          "metric": "catalog_munge",
          "count": 2,
          "mean": 94, 
          "aggregate": 188         
        },
        {
          "metric": "facts_save",                                                                        
          "count": 2,
          "mean": 76,
          "aggregate": 152
        },
        {
          "metric": "command_submit_replace_catalog",
          "count": 2,
          "mean": 54,
          "aggregate": 108
        },
        {
          "metric": "command_submit_replace_facts",
          "count": 2,
          "mean": 49,
          "aggregate": 98
        },
        {
          "metric": "report_process",
          "count": 2,
          "mean": 41,
          "aggregate": 82
        },
        {
          "metric": "command_submit_store_report",
          "count": 2,
          "mean": 30,
          "aggregate": 60
        },
        {
          "metric": "payload_format",
          "count": 8,
          "mean": 2,
          "aggregate": 16
        },
        {
          "metric": "report_convert_to_wire_format_hash",
          "count": 2,
          "mean": 4,
          "aggregate": 8
        },
        {
          "metric": "facts_encode",
          "count": 2,
          "mean": 2,
          "aggregate": 4
        }
      ],
      "inline-metrics": []
    }
  },
  "active_alerts": [],
  "service_name": "puppet-profiler"
}
```

Reply:
The updated set of metrics looks good to me 👍